### PR TITLE
main: Introduce "--" as option terminator

### LINKED ISSUE #1883: ctags -L: not possible to specifiy files with leading dashes (handled as option)
The man page for `-L file` says:

> Options are also accepted in this input.

This is a problem with real files with uncommon names like `-X`:

```
% printf '%s\n'  '-X' | ctags -L -
ctags: Unknown option: -X
```

I've tried passing in `--` in front to ignore filenames like `-X`, but this results in:

```
% printf '%s\n' '--' '-X' | ctags -L -
ctags: Unknown option: --
```

I think `-L -` should either not accept/handle options (backward incompatible?), or skip option parsing after `--`.

My use case is using `rg --files | ctags --links=no -L -` to generate tags.

### PR DESCRIPTION
**Pull request by masatake**
Close #1883.

To pass a file name started from "-" to the ctags command easier, this
change introduces "--" as option terminator in the command line
processing.

    $ cat ./-i
    #!/bin/sh
    function bar
    {
	:
    }
    $ ../../ctags -G -o - -i
    ctags: No files specified. Try "ctags --help".
    $ ../../ctags -G -o - ./-i
    bar	./-i	/^function bar$/;"	f
    $ ../../ctags -G -o - -- -i
    bar	-i	/^function bar$/;"	f

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

**Comment by masatake** [comment]
.... Removing  Tmain/nameless-long-option.d test case is not a good idea.
I will revert it partially.

**Comment by masatake** [comment]
"COMMAND LINE INTERFACE" in man/ctags.rst.1.in is the place to update.

File: man/ctags.1.rst.in [hunk 1]
```diff
@@ -533,16 +533,13 @@ are not listed here. They are experimental or for debugging purpose.
...
-	input. If this option is specified more than once, only the last will
-	apply. Note: file is read in line-oriented mode, where a new line is
-	the only delimiter and non-trailing white space is considered significant,
-	in order that file names containing spaces may be supplied
-	(however, trailing white space is stripped from lines); this can affect
-	how options are parsed if included in the input.
+	Read a list of file names and options from ``file`` (``-`` means
+	standard input).
+	File names read using this option are processed after file
+	names appearing on the command line.
```
Comment (blueyed) [inline man/ctags.1.rst.in]: Is this true for options also?
Would rephrase/simplify this then.

File: man/ctags.1.rst.in [hunk 2]
```diff
@@ -533,16 +533,13 @@ are not listed here. They are experimental or for debugging purpose.
...
+	Use ``--`` to handle following lines to be handled as file names only.
```
Comment (blueyed) [inline man/ctags.1.rst.in]: Needs a fixup (two times "handle").

Comment (blueyed) [reply man/ctags.1.rst.in]: Done.

**Comment by masatake** [comment]
How do you append a change to this pull request?
Very interesting technique.

"--" is not only for "-L -" option. "--" can be used anywhere on a command line.

I'm calling "--" "option terminator". Do you have any better name for it?

File: main/options.c [hunk 3]
```diff
@@ -871,15 +872,18 @@ static void cArgRead (cookedArgs *const current)
 		item = argItem (current->args);
 		current->shortOptions = NULL;
 		Assert (item != NULL);
-		if (strncmp (item, "--", (size_t) 2) == 0)
+
+		if ((!NonOptionEncountered)
+			&& (strncmp (item, "--", (size_t) 2) == 0))
```
Comment (b4n) [inline main/options.c]: Doesn't using the already existing `NonOptionEncountered` (which IIUC gets set to true when matching a non-option) break passing options after a file name altogether?  This would be unfortunate because for now it works fine, and there's even special handling for *some* options that should be passed before filenames.

Comment (masatake) [reply main/options.c]: @b4n, you are correct.
Furthermore, it seems that I broke the compatibility of the command line interface in the past.

Exuberant-ctags:
```
[jet@localhost]~/var/ctags% ctags  -o - foo.c --list-kinds=Sh          
f  functions
```

Univeresal-ctags:
```
[jet@localhost]~/var/ctags% ./ctags  -o - foo.c --list-kinds=Sh
ctags: Warning: cannot open input file "--list-kinds=Sh" : No such file or directory
a	foo.c	/^void a () {}$/;"	f	typeref:typename:void
b	foo.c	/^void b () {}$/;"	f	typeref:typename:void
```

U-ctags ignores the option (--list-kinds) after input file name.

Another example:
```
[jet@localhost]~/var/ctags% ctags  -o - foo.c --sort=no      
b	foo.c	/^void b () {}$/;"	f
a	foo.c	/^void a () {}$/;"	f
[jet@localhost]~/var/ctags% ./ctags  -o - foo.c --sort=no      
ctags: Warning: cannot open input file "--sort=no" : No such file or directory
a	foo.c	/^void a () {}$/;"	f	typeref:typename:void
b	foo.c	/^void b () {}$/;"	f	typeref:typename:void
```

This incompatibility is nothing to do with the newly introduced `--` option.
I will do bisect and spot the change that introduced the incompatibility first.
After that, I will revise this pull request.

Comment (masatake) [reply main/options.c]: I will work again on this item after reivising #1878.

**Comment by masatake** [comment]
What I should do first may be updating man/ctags-incompatibilities.7.rst.in.

**Comment by blueyed** [comment]
> How do you append a change to this pull request?

If the "allow maintainers to change branch" or similar checkbox is checked, others (with write permissions for the repo) can push there, too.

Sorry for the lack of feedback from me - just tried it again, when `rg --files` would include `-` in a project, used with `rg --files | --links=no -L -`.  This PR works there, but from quickly looking at the code  that might be handled as an option.

And `printf '-' | ./ctags --links=no -L -` still fails with "ctags: Unknown option: -".

**Comment by hirooih** [comment]
@masatake san,

> What I should do first may be updating man/ctags-incompatibilities.7.rst.in.

I've sent PR #2859 for this issue (#1889).

If you merge the fix of `--` option, we can close #596, #1883, and #1884.

**Comment by masatake** [comment]
@hirooih, thank you. Your change reduces the complexities behind this pull request.

After merging #2859, I can think about this issue without worrying about the incompatibilities. 
I will disallow passing an option via `-L -` option. In many places, I have assumed the values for options like ``--kinds-<LANG>=...`` are never changed during parsing files.

I will rework this item.
